### ISSUE TITLE
[Core] Ray may hang if workers fail to start due to limited ports

### ISSUE THREAD

**Issue by joshua-cogliati-inl** [ray-project/ray#28071]
### What happened + What you expected to happen

1. If `--min-worker-port` and `--max-worker-port` are used on a cluster, ray stops running any remote functions.
2. ray would be able to find a node that has an available port and port and run on it. (As in the sample given below, I am asking for 4 remotes to run on 6 nodes and have given each node 8 ports, and the run completely stops.) See #27499 for why I am using worker-ports. (From the Ray Architecture document: "Ray worker nodes are designed to be homogeneous, so that any single node may be lost without bringing down the entire cluster." Basically, this is a single node failure that stops progress in the entire cluster.)
3. below

Basically, ray status seems to think there are available cpus (1/6 used) but the ray_test python program is unable to find them.
```
ray status --address=10.159.0.79:58391
======== Autoscaler status: 2022-08-23 12:36:28.035691 ========
Node status
---------------------------------------------------------------
Healthy:
 1 node_991fc91accbb9cc5e9294c23d505be10da58c11a6f679c5562fab5b8
 1 node_c3238c760e1c81564e6d0dbff838b3678bba29bdd221c1dcaabfb19b
 1 node_a4b36956a2fed8fa17ca63ee3155b0c68bb2e04c094d5366eff0b9e5
 1 node_49970fd58f7f63a57aa2218156c24919dd6755b721a69260f2db7992
 1 node_50b4b421ab67619fd61ac967de3f72ebce9a3d1826755bd0c0b9ce86
 1 node_58f9c045ad2a1bda3ac013012ef758a00c05b93b88688051ca4c67d9
Pending:
 (no pending nodes)
Recent failures:
 (no failures)

Resources
---------------------------------------------------------------
Usage:
 1.0/6.0 CPU
 0.00/571.893 GiB memory
 0.00/249.089 GiB object_store_memory

Demands:
 {'CPU': 1.0}: 1+ pending tasks/actors
```

```
$ python ray_test.py
starting remote functions
remotes started
(raylet, ip=10.159.0.92) [2022-08-23 12:33:11,849 E 197968 197968] core_worker.cc:137: Failed to register worker 9389d63fbf6022579d3f4f76962f5e43aecd706984ed2535e68e4f88 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.94) [2022-08-23 12:33:48,045 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(55142) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-23 12:33:48,163 E 101832 101832] (raylet) worker_pool.cc:502: Some workers of the worker process(109680) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.92) [2022-08-23 12:34:10,844 E 190127 190127] (raylet) worker_pool.cc:502: Some workers of the worker process(197968) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.92) [2022-08-23 12:34:11,385 E 198185 198185] core_worker.cc:137: Failed to register worker c547d2a09340527eb29e316cbdbf34cdbb93fdda3dbdf8124902bc83 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.92) [2022-08-23 12:35:10,848 E 190127 190127] (raylet) worker_pool.cc:502: Some workers of the worker process(198185) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.92) [2022-08-23 12:35:11,634 E 198353 198353] core_worker.cc:137: Failed to register worker 8e47343dfaf4ddd666f2c9b7e1767fc822b09a6e908d575e620e8df8 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
...
```

```
#running ps -u fred -o pcpu,pid,ppid,thcount,user,size,stat,comm,args | grep 'ray::'
#start: 
93.1  94252  86552    21 fred   7812360 Rl ray::f()        ray::f()
90.6 159525 152200    21 fred   7840232 Rl ray::f()        ray::f()
96.3 136620 122245    21 fred   1914708 Rl ray::f()        ray::f()
#when making no progress
23.5 136620 122245    21 fred   1398696 Sl ray::IDLE       ray::IDLE
23.4  94252  86552    21 fred   7357788 Sl ray::IDLE       ray::IDLE
23.3 159525 152200    21 fred   7357852 Sl ray::IDLE       ray::IDLE
```

### Versions / Dependencies

(Note that this was tested with ray 2.0 and still exists: https://github.com/ray-project/ray/issues/28071#issuecomment-1227623743  )

```
$ conda list
# packages in environment at /home/fred/miniconda3/envs/raven_libraries_heron_newer_ray:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
_tflow_select             2.3.0                       mkl  
absl-py                   0.15.0             pyhd8ed1ab_0    conda-forge
aiohttp                   3.8.1            py37h540881e_1    conda-forge
aiohttp-cors              0.7.0                    pypi_0    pypi
aiosignal                 1.2.0              pyhd8ed1ab_0    conda-forge
ampl-mp                   3.1.0             h2cc385e_1006    conda-forge
astor                     0.8.1              pyh9f0ad1d_0    conda-forge
astroid                   2.11.6           py37h89c1867_0    conda-forge
async-timeout             4.0.2              pyhd8ed1ab_0    conda-forge
asynctest                 0.13.0                     py_0    conda-forge
attrs                     22.1.0             pyh71513ae_1    conda-forge
blessed                   1.19.1                   pypi_0    pypi
blinker                   1.4                        py_1    conda-forge
brotlipy                  0.7.0           py37h540881e_1004    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2022.6.15            ha878542_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
cachetools                5.2.0              pyhd8ed1ab_0    conda-forge
certifi                   2022.6.15        py37h89c1867_0    conda-forge
cffi                      1.15.1           py37h43b0acd_0    conda-forge
cftime                    1.6.0            py37h6c7ee08_0    conda-forge
charset-normalizer        2.1.0              pyhd8ed1ab_0    conda-forge
click                     8.0.4                    pypi_0    pypi
cloudpickle               1.6.0                      py_0    conda-forge
coin-or-cbc               2.10.8               h3786ebc_0    conda-forge
coin-or-cgl               0.60.6               h6f57e76_1    conda-forge
coin-or-clp               1.17.7               hc56784d_1    conda-forge
coin-or-osi               0.108.7              h2720bb7_1    conda-forge
coin-or-utils             2.11.6               h202d8b1_1    conda-forge
coincbc                   2.10.8            0_metapackage    conda-forge
colorama                  0.4.5              pyhd8ed1ab_0    conda-forge
colorful                  0.5.4                    pypi_0    pypi
coverage                  6.4.4            py37h540881e_0    conda-forge
cryptography              37.0.4           py37h38fbfac_0    conda-forge
curl                      7.83.1               h7bff187_0    conda-forge
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
dbus                      1.13.6               h5008d03_3    conda-forge
dill                      0.3.5.1            pyhd8ed1ab_0    conda-forge
distlib                   0.3.5                    pypi_0    pypi
elementpath               3.0.2                    pypi_0    pypi
expat                     2.4.8                h27087fc_0    conda-forge
filelock                  3.8.0                    pypi_0    pypi
fontconfig                2.14.0               h8e229c2_0    conda-forge
freetype                  2.12.1               hca18f0e_0    conda-forge
frozenlist                1.3.1            py37h540881e_0    conda-forge
gast                      0.2.2                      py_0    conda-forge
gettext                   0.19.8.1          h73d1719_1008    conda-forge
glib                      2.72.1               h6239696_0    conda-forge
glib-tools                2.72.1               h6239696_0    conda-forge
glpk                      5.0                  h445213a_0    conda-forge
gmp                       6.2.1                h58526e2_0    conda-forge
google-api-core           2.8.2                    pypi_0    pypi
google-auth               2.10.0             pyh6c4a22f_0    conda-forge
google-auth-oauthlib      0.4.6              pyhd8ed1ab_0    conda-forge
google-pasta              0.2.0              pyh8c360ce_0    conda-forge
googleapis-common-protos  1.56.4                   pypi_0    pypi
gpustat                   1.0.0rc1                 pypi_0    pypi
grpc-cpp                  1.48.0               hbad87ad_3    conda-forge
grpcio                    1.43.0                   pypi_0    pypi
gst-plugins-base          1.14.5               h0935bb2_2    conda-forge
gstreamer                 1.18.5               h9f60fe5_3    conda-forge
h5py                      3.6.0           nompi_py37hd308b1e_100    conda-forge
hdf4                      4.2.15               h9772cbc_4    conda-forge
hdf5                      1.12.1          nompi_h2386368_104    conda-forge
icu                       67.1                 he1b5a44_0    conda-forge
idna                      3.3                pyhd8ed1ab_0    conda-forge
imageio                   2.9.0                      py_0    conda-forge
importlib-metadata        4.11.4           py37h89c1867_0    conda-forge
importlib-resources       5.9.0                    pypi_0    pypi
importlib_metadata        4.11.4               hd8ed1ab_0    conda-forge
ipopt                     3.14.9               hc8a599a_0    conda-forge
isort                     5.10.1             pyhd8ed1ab_0    conda-forge
joblib                    1.1.0              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   h166bdaf_2    conda-forge
jsonschema                4.10.0                   pypi_0    pypi
keras-applications        1.0.8                      py_1    conda-forge
keras-preprocessing       1.1.2              pyhd8ed1ab_0    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
kiwisolver                1.4.4            py37h7cecad7_0    conda-forge
krb5                      1.19.3               h3790be6_0    conda-forge
lazy-object-proxy         1.7.1            py37h540881e_1    conda-forge
lcms2                     2.12                 hddcbb42_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
lerc                      4.0.0                h27087fc_0    conda-forge
libabseil                 20220623.0      cxx17_h48a1fff_1    conda-forge
libblas                   3.9.0           16_linux64_openblas    conda-forge
libcblas                  3.9.0           16_linux64_openblas    conda-forge
libclang                  11.1.0          default_ha53f305_1    conda-forge
libcurl                   7.83.1               h7bff187_0    conda-forge
libdeflate                1.13                 h166bdaf_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libevent                  2.1.10               h9b69904_4    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 12.1.0              h8d9b700_16    conda-forge
libgfortran-ng            12.1.0              h69a702a_16    conda-forge
libgfortran5              12.1.0              hdcd56e2_16    conda-forge
libglib                   2.72.1               h2d90d5f_0    conda-forge
libgomp                   12.1.0              h8d9b700_16    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
liblapack                 3.9.0           16_linux64_openblas    conda-forge
liblapacke                3.9.0           16_linux64_openblas    conda-forge
libllvm11                 11.1.0               hf817b99_3    conda-forge
libnetcdf                 4.8.1           nompi_h329d8a1_102    conda-forge
libnghttp2                1.47.0               hdcd2b5c_1    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libopenblas               0.3.21          pthreads_h78a6416_1    conda-forge
libpng                    1.6.37               h753d276_4    conda-forge
libpq                     12.9                 h16c4e8d_3  
libprotobuf               3.20.1               h6239696_1    conda-forge
libsqlite                 3.39.2               h753d276_1    conda-forge
libssh2                   1.10.0               haa6b8db_3    conda-forge
libstdcxx-ng              12.1.0              ha89aaad_16    conda-forge
libtiff                   4.4.0                h0e0dad5_3    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libwebp-base              1.2.4                h166bdaf_0    conda-forge
libxcb                    1.13              h7f98852_1004    conda-forge
libxkbcommon              1.0.3                he3ba5ed_0    conda-forge
libxml2                   2.9.10               h68273f3_2    conda-forge
libxslt                   1.1.33               hf705e74_1    conda-forge
libzip                    1.9.2                hc869a4a_1    conda-forge
libzlib                   1.2.12               h166bdaf_2    conda-forge
lxml                      4.8.0            py37h540881e_3    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
markdown                  3.4.1              pyhd8ed1ab_0    conda-forge
matplotlib                3.2.2                         1    conda-forge
matplotlib-base           3.2.2            py37h1d35a4c_1    conda-forge
mccabe                    0.7.0              pyhd8ed1ab_0    conda-forge
metis                     5.1.0             h58526e2_1006    conda-forge
msgpack                   1.0.4                    pypi_0    pypi
multidict                 6.0.2            py37h540881e_1    conda-forge
mumps-include             5.2.1               ha770c72_11    conda-forge
mumps-seq                 5.2.1               h2104b81_11    conda-forge
mysql-common              8.0.30               haf5c9bc_0    conda-forge
mysql-libs                8.0.30               h28c427c_0    conda-forge
ncurses                   6.3                  h27087fc_1    conda-forge
netcdf4                   1.5.8           nompi_py37hf784469_101    conda-forge
nomkl                     1.0                  h5ca1d4c_0    conda-forge
nose                      1.3.7                   py_1006    conda-forge
nspr                      4.32                 h9c3ff4c_1    conda-forge
nss                       3.78                 h2350873_0    conda-forge
numexpr                   2.8.0           py37hfe5f03c_101    conda-forge
numpy                     1.18.5           py37h8960a57_0    conda-forge
numpy-financial           1.0.0              pyhd8ed1ab_0    conda-forge
nvidia-ml-py              11.495.46                pypi_0    pypi
oauthlib                  3.2.0              pyhd8ed1ab_0    conda-forge
opencensus                0.11.0                   pypi_0    pypi
opencensus-context        0.1.3                    pypi_0    pypi
openjpeg                  2.5.0                h7d73246_1    conda-forge
openssl                   1.1.1q               h166bdaf_0    conda-forge
opt_einsum                3.3.0              pyhd8ed1ab_1    conda-forge
pandas                    1.1.5            py37hdc94413_0    conda-forge
patsy                     0.5.2              pyhd8ed1ab_0    conda-forge
pcre                      8.45                 h9c3ff4c_0    conda-forge
pillow                    9.2.0            py37h850a105_2    conda-forge
pip                       22.2.2             pyhd8ed1ab_0    conda-forge
pkgutil-resolve-name      1.3.10                   pypi_0    pypi
platformdirs              2.5.2              pyhd8ed1ab_1    conda-forge
ply                       3.11                       py_1    conda-forge
prometheus-client         0.13.1                   pypi_0    pypi
protobuf                  3.20.1           py37hd23a5d3_0    conda-forge
psutil                    5.9.1            py37h540881e_0    conda-forge
pthread-stubs             0.4               h36c2ea0_1001    conda-forge
py-spy                    0.3.12                   pypi_0    pypi
pyasn1                    0.4.8                      py_0    conda-forge
pyasn1-modules            0.2.7                      py_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyjwt                     2.4.0              pyhd8ed1ab_0    conda-forge
pylint                    2.14.5             pyhd8ed1ab_0    conda-forge
pyomo                     6.4.2            py37hd23a5d3_0    conda-forge
pyopenssl                 22.0.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pyrsistent                0.18.1                   pypi_0    pypi
pyside2                   5.13.2           py37hfa98aef_4    conda-forge
pysocks                   1.7.1            py37h89c1867_5    conda-forge
python                    3.7.12          hb7a2778_100_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.7                     2_cp37m    conda-forge
pytz                      2022.2.1           pyhd8ed1ab_0    conda-forge
pyu2f                     0.1.5              pyhd8ed1ab_0    conda-forge
pyutilib                  6.0.0              pyh9f0ad1d_0    conda-forge
pyyaml                    6.0                      pypi_0    pypi
qt                        5.12.9               h763d07f_1    conda-forge
ray                       1.13.0                   pypi_0    pypi
re2                       2022.06.01           h27087fc_0    conda-forge
readline                  8.1.2                h0f457ee_0    conda-forge
requests                  2.28.1             pyhd8ed1ab_0    conda-forge
requests-oauthlib         1.3.1              pyhd8ed1ab_0    conda-forge
rsa                       4.9                pyhd8ed1ab_0    conda-forge
scikit-learn              0.24.2           py37hf0f1638_1    conda-forge
scipy                     1.5.3            py37h14a347d_0    conda-forge
scotch                    6.0.9                hb2e6521_2    conda-forge
setuptools                59.8.0           py37h89c1867_1    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
smart-open                6.0.0                    pypi_0    pypi
sqlite                    3.39.2               h4ff8645_1    conda-forge
statsmodels               0.12.2           py37hb1e94ed_0    conda-forge
swig                      4.0.2                hd3c618e_2    conda-forge
tensorboard               2.8.0              pyhd8ed1ab_1    conda-forge
tensorboard-data-server   0.6.0            py37h38fbfac_2    conda-forge
tensorboard-plugin-wit    1.8.1              pyhd8ed1ab_0    conda-forge
tensorflow                2.0.0           mkl_py37h66b46cc_0  
tensorflow-base           2.0.0           mkl_py37h9204916_0  
tensorflow-estimator      2.5.0              pyh8a188c0_0    conda-forge
termcolor                 1.1.0              pyhd8ed1ab_3    conda-forge
threadpoolctl             3.1.0              pyh8a188c0_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
tomlkit                   0.11.4             pyha770c72_0    conda-forge
tornado                   6.2              py37h540881e_0    conda-forge
typed-ast                 1.5.4            py37h540881e_0    conda-forge
typing                    3.10.0.0           pyhd8ed1ab_0    conda-forge
typing-extensions         4.3.0                hd8ed1ab_0    conda-forge
typing_extensions         4.3.0              pyha770c72_0    conda-forge
unixodbc                  2.3.10               h583eb01_0    conda-forge
urllib3                   1.26.11            pyhd8ed1ab_0    conda-forge
virtualenv                20.16.3                  pypi_0    pypi
wcwidth                   0.2.5                    pypi_0    pypi
werkzeug                  0.16.1                     py_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
wrapt                     1.14.1           py37h540881e_0    conda-forge
xarray                    0.16.2             pyhd8ed1ab_0    conda-forge
xmlschema                 2.0.2                    pypi_0    pypi
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
yarl                      1.7.2            py37h540881e_2    conda-forge
zipp                      3.8.1              pyhd8ed1ab_0    conda-forge
zlib                      1.2.12               h166bdaf_2    conda-forge
zstd                      1.5.2                h8a70e8d_4    conda-forge

$ pip list
Package                  Version
------------------------ ---------
absl-py                  0.15.0
aiohttp                  3.8.1
aiohttp-cors             0.7.0
aiosignal                1.2.0
astor                    0.8.1
astroid                  2.11.6
async-timeout            4.0.2
asynctest                0.13.0
attrs                    22.1.0
blessed                  1.19.1
blinker                  1.4
brotlipy                 0.7.0
cached-property          1.5.2
cachetools               5.2.0
certifi                  2022.6.15
cffi                     1.15.1
cftime                   1.6.0
charset-normalizer       2.1.0
click                    8.0.4
cloudpickle              1.6.0
colorama                 0.4.5
colorful                 0.5.4
coverage                 6.4.4
cryptography             37.0.4
cycler                   0.11.0
DaemonLite               0.0.2
dill                     0.3.5.1
distlib                  0.3.5
elementpath              3.0.2
filelock                 3.8.0
frozenlist               1.3.1
gast                     0.2.2
google-api-core          2.8.2
google-auth              2.10.0
google-auth-oauthlib     0.4.6
google-pasta             0.2.0
googleapis-common-protos 1.56.4
gpustat                  1.0.0rc1
grpcio                   1.43.0
h5py                     3.6.0
h5py-wrapper             1.1.0
hdfdict                  0.3.1
idna                     3.3
imageio                  2.9.0
importlib-metadata       4.11.4
importlib-resources      5.9.0
isort                    5.10.1
joblib                   1.1.0
jsonschema               4.10.0
Keras-Applications       1.0.8
Keras-Preprocessing      1.1.2
kiwisolver               1.4.4
lazy-object-proxy        1.7.1
lxml                     4.8.0
Markdown                 3.4.1
matplotlib               3.2.2
mccabe                   0.7.0
msgpack                  1.0.4
multidict                6.0.2
netCDF4                  1.5.8
nose                     1.3.7
numexpr                  2.8.0
numpy                    1.18.5
numpy-financial          1.0.0
nvidia-ml-py             11.495.46
oauthlib                 3.2.0
opencensus               0.11.0
opencensus-context       0.1.3
opt-einsum               3.3.0
pandas                   1.1.5
patsy                    0.5.2
Pillow                   9.2.0
pip                      22.2.2
pkgutil_resolve_name     1.3.10
platformdirs             2.5.2
ply                      3.11
prometheus-client        0.13.1
protobuf                 3.20.1
psutil                   5.9.1
py-spy                   0.3.12
pyasn1                   0.4.8
pyasn1-modules           0.2.7
pycparser                2.21
PyJWT                    2.4.0
pylint                   2.14.5
Pyomo                    6.4.2
pyOpenSSL                22.0.0
pyparsing                3.0.9
pyrsistent               0.18.1
PySocks                  1.7.1
pytest-runner            5.3.1
python-dateutil          2.8.2
pytz                     2022.2.1
pyu2f                    0.1.5
PyUtilib                 6.0.0
PyYAML                   6.0
ray                      1.13.0
requests                 2.28.1
requests-oauthlib        1.3.1
rsa                      4.9
scikit-learn             0.24.2
scipy                    1.5.3
setuptools               59.8.0
six                      1.16.0
smart-open               6.0.0
statsmodels              0.12.2
tensorboard              2.8.0
tensorboard-data-server  0.6.0
tensorboard-plugin-wit   1.8.1
tensorflow               2.0.0
tensorflow-estimator     2.5.0
termcolor                1.1.0
threadpoolctl            3.1.0
tomli                    2.0.1
tomlkit                  0.11.4
tornado                  6.2
typed-ast                1.5.4
typing_extensions        4.3.0
urllib3                  1.26.11
virtualenv               20.16.3
wcwidth                  0.2.5
Werkzeug                 0.16.1
wheel                    0.37.1
wrapt                    1.14.1
xarray                   0.16.2
xmlschema                2.0.2
xmltodict                0.12.0
yarl                     1.7.2
zipp                     3.8.1
$ python
Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:53) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

### Reproduction script

Scripts used:
creation of environment:
```bash
conda install --name raven_libraries_heron_newer_ray -y -c conda-forge h5py numpy=1.18 scipy=1.5 scikit-learn=0.24 pandas=1.1 xarray=0.16 netcdf4=1.5 matplotlib=3.2 statsmodels=0.12 cloudpickle=1.6 tensorflow=2.0 python=3 hdf5 swig pylint coverage lxml psutil pip importlib_metadata pyside2 nomkl numexpr imageio=2.9 setuptools dill pyomo pyutilib glpk ipopt coincbc numpy-financial
pip install  ray[default]==1.13.* xmlschema
```

ray_start.sh
```bash
#!/bin/bash                                                                     

source /home/fred/miniconda3/etc/profile.d/conda.sh
conda activate raven_libraries_heron_newer_ray

NUM_CPUS=$1
HEAD_ADDRESS=$2

ray start --verbose --address=$HEAD_ADDRESS --num-cpus $NUM_CPUS  --min-worker-port 10002 --max-worker-port $((10002+8*$NUM_CPUS))
```

ray_test.py
```python
import ray
import time
ray.init(address='auto')

start = time.time()

@ray.remote
def f(x):
    a = [2**2**2**2**2 for x in range(100000)]
    return x * x

print("starting remote functions")
futures = [f.remote(i) for i in range(4)]
print("remotes started")
print(ray.get(futures)) # [0, 1, 4, 9]                                          
end = time.time()
print("time: ",end - start)
```

Head node:
```bash
ray start --head --num-cpus=1 --port=0 --min-worker-port 10002 --max-worker-port 10010`
...
#adjust RAY_ADDRESS based on output of above
export RAY_ADDRESS=10.159.0.79:60085

#Note that PBS_NODEFILE is a list of nodes hostnames that can be used
for NODE in `cat $PBS_NODEFILE | uniq`; do
    if echo $NODE | grep -q `hostname`; then
        echo skipping $NODE;
    else
       echo $NODE
       ssh $NODE /home/fred/notes-for-various-projects/raven/ray_start.sh 1 $RAY_ADDRESS
    fi
done

python ray_test.py
```

I checked that ports were available by running: `netstat --all | grep ":1000"`

### Issue Severity

High: It blocks me from completing my task.

**Comment by joshua-cogliati-inl** [comment]
Note that there does not seem to be a way to get the 'ray::IDLE' to finish.
For example, I delete the futures after getting the data:
```python
import ray
import time
import gc
ray.init(address='auto')

start = time.time()

@ray.remote
def f(x):
    a = [2**2**2**2**2 for x in range(100000)]
    return x * x

print("starting remote functions")
futures = [f.remote(i) for i in range(4)]
print("remotes started")
left = len(futures)
outputs = [None]*left
while left > 0:
    for i in range(len(futures)):
        if futures[i] is not None:
            try:
                f = futures[i]
                runReturn = ray.get(f, timeout=1e-10)
                print(i, runReturn)
                left += -1
                outputs[i] = runReturn
                futures[i] = None
                gc.collect()
                print("referrers", len(gc.get_referrers(f)))
                del f
                gc.collect()
            except ray.exceptions.GetTimeoutError:
                pass

print(outputs) # [0, 1, 4, 9]
end = time.time()
print("time: ",end - start)
```
and get this output:
```
$ python ray_test2.py
starting remote functions
remotes started
(raylet, ip=10.159.0.94) [2022-08-23 13:32:10,977 E 65841 65841] core_worker.cc:137: Failed to register worker 56f00aa1786b6ff744315fab7ab4e9732b2b89933205b7871afc5e2c to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
1 1
referrers 1
0 0
referrers 1
2 4
referrers 1
(raylet, ip=10.159.0.92) [2022-08-23 13:32:50,658 E 190127 190127] (raylet) worker_pool.cc:502: Some workers of the worker process(208990) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-23 13:32:50,658 E 101832 101832] (raylet) worker_pool.cc:502: Some workers of the worker process(119671) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:33:10,149 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(65841) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:33:10,696 E 66041 66041] core_worker.cc:137: Failed to register worker 8505d9d63a607589f1e3a2ec15e623f3bc86746e9e9a90fa0591169c to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.94) [2022-08-23 13:34:10,150 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(66041) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:34:10,968 E 66205 66205] core_worker.cc:137: Failed to register worker 83db274ec9958e958277d71243c019be1c42297c4e52c9201e9b844d to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.94) [2022-08-23 13:35:10,151 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(66205) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:35:10,738 E 66354 66354] core_worker.cc:137: Failed to register worker 0e216954d81fd80fb8830c065821895400b8a8fb553b3f6c1dedef76 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
```
and this ps grep 'ray::' output:
```
43.9 147723 122245    21 fred   1398820 Sl ray::IDLE       ray::IDLE
43.5 103671  86552    21 fred   7357716 Sl ray::IDLE       ray::IDLE
44.0 169080 152200    21 fred   7357788 Sl ray::IDLE       ray::IDLE
```
As in there is only one remote waiting to complete, but ray is unable to complete it because it is blocked by ray futures that have already returned and been deleted.

**Comment by joshua-cogliati-inl** [comment]
Oddly enough, if I increase the range to 10 `futures = [f.remote(i) for i in range(10)]`
it is able to run a second batch of 3, but it still fails to finish.

```
$ python ray_test2.py
starting remote functions
remotes started
(raylet, ip=10.159.0.92) [2022-08-23 13:43:56,522 E 211007 211007] core_worker.cc:137: Failed to register worker 60d50bcb76025d09e144ccf0e80922c9613bab2b5e14f940f1aeccfe to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.94) [2022-08-23 13:43:56,529 E 67951 67951] core_worker.cc:137: Failed to register worker 476f27c6cf38faccfac69d01e9bca324935ca002aac6b9c68ac39d55 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-23 13:43:56,530 E 121634 121634] core_worker.cc:137: Failed to register worker ed0af77bdd4b5c3e3fddf6c4f4ca4cbb1028b692c2e521bee7fbf00b to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(scheduler +4s) Tip: use `ray status` to view detailed cluster status. To disable these messages, set RAY_SCHEDULER_EVENTS=0.
(scheduler +4s) Warning: The following resource request cannot be scheduled right now: {'CPU': 1.0}. This is likely due to all cluster resources being claimed by actors. Consider creating fewer actors or adding more nodes to this Ray cluster.
1 1
referrers 1
0 0
referrers 1
2 4
referrers 1
(scheduler +39s) Warning: The following resource request cannot be scheduled right now: {'CPU': 1.0}. This is likely due to all cluster resources being claimed by actors. Consider creating fewer actors or adding more nodes to this Ray cluster.
(raylet, ip=10.159.0.92) [2022-08-23 13:44:55,714 E 190127 190127] (raylet) worker_pool.cc:502: Some workers of the worker process(211007) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:44:55,710 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(67951) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-23 13:44:55,714 E 101832 101832] (raylet) worker_pool.cc:502: Some workers of the worker process(121634) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:44:56,289 E 68103 68103] core_worker.cc:137: Failed to register worker eb1c86f97c1614d8a5fef54c33df3ee3266b5d64e9dc6678906c7fc7 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.92) [2022-08-23 13:44:56,344 E 211155 211155] core_worker.cc:137: Failed to register worker 3f504ac4bf1b954709f43758befd8b16af3b5e06a9041b3a4e1ec72b to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-23 13:44:56,540 E 121785 121785] core_worker.cc:137: Failed to register worker dcbf45970b2e391d30201a08aa7e7d9732c1c5c7cff5a6fd2e52a20a to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
4 16
referrers 1
5 25
referrers 1
3 9
referrers 1
6 36
referrers 1
(raylet, ip=10.159.0.81) [2022-08-23 13:45:55,717 E 101832 101832] (raylet) worker_pool.cc:502: Some workers of the worker process(121785) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:45:55,711 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(68103) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.92) [2022-08-23 13:45:55,717 E 190127 190127] (raylet) worker_pool.cc:502: Some workers of the worker process(211155) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-23 13:45:56,377 E 121947 121947] core_worker.cc:137: Failed to register worker a1070438c5b95a02ae159158ac0bd76ec5169b77d6d73c17f52cb622 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.94) [2022-08-23 13:45:56,520 E 68389 68389] core_worker.cc:137: Failed to register worker c3cc324f5fdb97a4d3fc37a3a59cd767ae17b4c48d7c5a680f767db2 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.92) [2022-08-23 13:45:56,545 E 211319 211319] core_worker.cc:137: Failed to register worker 10b7fa9c8a0df4a26a3439dacdc27db701e89d97f942bdf999328a97 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-23 13:46:55,721 E 101832 101832] (raylet) worker_pool.cc:502: Some workers of the worker process(121947) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:46:55,712 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(68389) have not registered within the timeout. The process is dead, probably it crashed during start.
```

**Comment by joshua-cogliati-inl** [comment]
I tried with `futures = [f.remote(i) for i in range(20)]` and it finished the first 16:
```
(raylet, ip=10.159.0.94) [2022-08-23 13:57:56,354 E 70936 70936] core_worker.cc:137: Failed to register worker c52534455be3591d7cdfb37996bd64fd5dc29dcee6b6494b9821ec97 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.92) [2022-08-23 13:57:56,418 E 213858 213858] core_worker.cc:137: Failed to register worker 675bf19406f39236641806f363bb153cf018b0eda1a51ea942806c14 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-23 13:57:56,429 E 124260 124260] core_worker.cc:137: Failed to register worker 78e0c0467ed76e3498943a15f8825b310df2187fbc50a533eb679614 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
9 81
referrers 1
10 100
referrers 1
11 121
referrers 1
12 144
referrers 1
13 169
referrers 1
14 196
referrers 1
15 225
referrers 1
16 256
referrers 1
(raylet, ip=10.159.0.81) [2022-08-23 13:58:55,816 E 101832 101832] (raylet) worker_pool.cc:502: Some workers of the worker process(124260) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.94) [2022-08-23 13:58:55,807 E 46719 46719] (raylet) worker_pool.cc:502: Some workers of the worker process(70936) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.92) [2022-08-23 13:58:55,816 E 190127 190127] (raylet) worker_pool.cc:502: Some workers of the worker process(213858) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-23 13:58:56,621 E 124405 124405] core_worker.cc:137: Failed to register worker 402b0d4b26169ddfd55c1bfa6d4556edb665e154bb8af37b7053120c to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.94) [2022-08-23 13:58:56,642 E 71115 71115] core_worker.cc:137: Failed to register worker f2a39c3b732027e852cf49e5eaf03f3b8fa9e5f3f99fd52509ac6241 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.

```
but then eventually died with:
```
(raylet, ip=10.159.0.94) [2022-08-23 14:45:56,679 E 80360 80360] core_worker.cc:137: Failed to register worker 22fd77568611516809ae76a918157ce0ef9277a4c9bd7b6b9d501483 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.92) [2022-08-23 14:45:56,822 E 223229 223229] core_worker.cc:137: Failed to register worker 075257de0b4900f8ef99858533332e7b364686f273bbe23996542dde to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-23 14:45:56,808 E 133327 133327] core_worker.cc:137: Failed to register worker fd58c7b27bbc15564b0f96cccc6ee4498a5626ef163657d1c402e1dd to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
[2022-08-23 14:47:19,065 C 152347 152374] gcs_client.cc:300: Couldn't reconnect to GCS server. The last attempted GCS server address was 10.159.0.79:57127
*** StackTrace Information ***
    ray::SpdLogMessage::Flush()
    ray::RayLog::~RayLog()
    ray::gcs::GcsClient::ReconnectGcsServer()
    ray::gcs::GcsClient::PeriodicallyCheckGcsConnection()
    ray::PeriodicalRunner::DoRunFnPeriodicallyInstrumented()
    std::_Function_handler<>::_M_invoke()
    EventTracker::RecordExecution()
    ray::PeriodicalRunner::DoRunFnPeriodicallyInstrumented()::{lambda()#1}::operator()()
    boost::asio::detail::wait_handler<>::do_complete()
    boost::asio::detail::scheduler::do_run_one()
    boost::asio::detail::scheduler::run()
    boost::asio::io_context::run()
    ray::core::CoreWorker::RunIOService()
    execute_native_thread_routine
```
Oddly enough, I was able to run a different ray_test.py (with size 4) at the same time that finished.

**Comment by joshua-cogliati-inl** [comment]
`ray start --head --num-cpus=2 --port=0 --min-worker-port 10002 --max-worker-port 10010` without running ray start on remote nodes works fine.

`ray start --head --num-cpus=4 --port=0 --min-worker-port 12002 --max-worker-port 12003` had a lot of warnings but finished:
```
$ export RAY_ADDRESS='10.159.0.79:48913'
$ python ray_test2.py                                                               starting remote functions
remotes started
(raylet) [2022-08-23 15:13:15,557 E 223585 223585] core_worker.cc:137: Failed to register worker a20a041a81c6dc5a5709d7e342bd95bec33fd5b2b181fcce5a4d9221 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:13:15,608 E 223583 223583] core_worker.cc:137: Failed to
 register worker dc970c6758178e2a45716b6d92deae5e236542eb8e8617bf8627cb7c to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:13:15,612 E 223586 223586] core_worker.cc:137: Failed to register worker 08a4a3e71092b95fa5364fe4708e920b17dc91978bc9b6edbc800ad3 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:13:16,706 E 223656 223656] core_worker.cc:137: Failed to register worker d1fcd9e5c8bcac1b9dc0bef3c3a88972e86640b807a650e252761382 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
0 0
referrers 1
1 1
referrers 1
(raylet) [2022-08-23 15:14:14,368 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(223583) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:14:14,369 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(223585) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:14:14,376 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(223586) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:14:15,657 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(223656) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:14:15,772 E 223905 223905] core_worker.cc:137: Failed to register worker ae77a12ec25a574ae0365b453d61678ba2ae16c4fa909111c2485781 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:14:15,784 E 223906 223906] core_worker.cc:137: Failed to register worker bbb31b2cea337d354c13129bc5e68ff84fb9c4bff6375acf68407d42 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:14:16,775 E 223914 223914] core_worker.cc:137: Failed to register worker 8449cdf67b13743211113cb433a152af730fd9f1784061a6ab5ca595 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
2 4
referrers 1
3 9
referrers 1
4 16
referrers 1
(raylet) [2022-08-23 15:15:14,371 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(223905) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:15:14,377 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(223906) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:15:15,373 E 224167 224167] core_worker.cc:137: Failed to register worker 37335b63b1d92787fe1cbed0d6c5faa1058e33272b888cc129afa913 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:15:16,827 E 224202 224202] core_worker.cc:137: Failed to register worker 3a86e3d7a43f55a6b8f356fbbe762df3afeac226ae0aa4c6c3a525cb to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
5 25
referrers 1
6 36
referrers 1
(raylet) [2022-08-23 15:16:14,373 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(224167) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:16:14,380 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(224168) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:16:15,523 E 224431 224431] core_worker.cc:137: Failed to register worker 8432e4cabb69fde410c9223885fb6c43c56cd72b35941def3a4b6c64 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet) [2022-08-23 15:16:15,669 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(224202) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:16:16,717 E 224452 224452] core_worker.cc:137: Failed to register worker c217c80628ecdae93504b1962213171116a47a0f57d60424d2230a5f to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
7 49
referrers 1
(raylet) [2022-08-23 15:17:14,397 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(224431) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:17:15,681 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(224452) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet) [2022-08-23 15:17:16,890 E 224689 224689] core_worker.cc:137: Failed to register worker b056087656ac3b635c04b8c0f02fc9803f80889646e8da159a4b197e to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
 64
referrers 1
(raylet) [2022-08-23 15:18:15,682 E 223289 223289] (raylet) worker_pool.cc:502: Some workers of the worker process(224689) have not registered within the timeout. The process is dead, probably it crashed during start.
9 81
referrers 1
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
time:  321.66390562057495

```
so the problem seems to only occur when using remote ray running.

**Comment by joshua-cogliati-inl** [comment]
I think this might be more a bug of robustness.  As in I can completely block ray from completing tasks by blocking one node from running workers, even with there being other nodes that are running fine.  For example, if I run this (listen_range.py) on one node:
```python
import socket
import time

l = []
for PORT in range(10002, 10010+1):
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.bind(('', PORT))
        s.listen(1)
        l.append(s)
    except OSError:
        print("unable to get ",PORT)

if len(l) > 0:
    time.sleep(600)
```
then it blocks the above ray_test2.py from finishing until the sleep(600) finishes, even with there being available raylets with workers on other nodes.

Output of ray_test2 (range 10) while running listen_range.py:
```
$ python ray_test2.py
starting remote functions
remotes started
(raylet, ip=10.159.0.81) [2022-08-24 09:02:57,650 E 22814 22814] core_worker.cc:137: Failed to register worker 8b8dd67cd10654be53e1445a598a750d637b6969e9008429076ceaba to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(scheduler +2s) Tip: use `ray status` to view detailed cluster status. To disable these messages, set RAY_SCHEDULER_EVENTS=0.
(scheduler +2s) Warning: The following resource request cannot be scheduled right now: {'CPU': 1.0}. This is likely due to all cluster resources being claimed by actors. Consider creating fewer actors or adding more nodes to this Ray cluster.
1 1
referrers 1
0 0
referrers 1
2 4
referrers 1
3 9
referrers 1
4 16
referrers 1
5 25
referrers 1
(raylet, ip=10.159.0.81) [2022-08-24 09:03:56,816 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(22814) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:03:57,742 E 23134 23134] core_worker.cc:137: Failed to register worker aa01a865bdb7b0818a899de18e18cf8162556c7ea1c4ce02768ace6f to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
8 64
referrers 1
7 49
referrers 1
6 36
referrers 1
(raylet, ip=10.159.0.81) [2022-08-24 09:04:56,821 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(23134) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:04:57,727 E 23321 23321] core_worker.cc:137: Failed to register worker d9bab59cebc6623b6888e5f9cba6756538880a24e452145b9b47bbbe to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:05:56,825 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(23321) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:05:57,766 E 23506 23506] core_worker.cc:137: Failed to register worker 07831a1b823b4d0833c7bacbe4203089b0451228d858fbaa21ba3c97 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:06:56,829 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(23506) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:06:57,683 E 23696 23696] core_worker.cc:137: Failed to register worker 21ec808ed9daef6f6564a8c511da61a2c90f04dd176f7f80f5a02709 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:07:56,833 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(23696) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:07:57,525 E 23920 23920] core_worker.cc:137: Failed to register worker 698e35848adee00ed063724dc9f5dc10db031de606f445a1392d9f48 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:08:56,837 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(23920) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:08:57,764 E 24107 24107] core_worker.cc:137: Failed to register worker 342d9a99b9ada6173f651a97d177d7305a0294baa04809662edecc54 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:09:56,841 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(24107) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:09:57,790 E 24283 24283] core_worker.cc:137: Failed to register worker 04977d09a86178cd50adb5c0ea058df44fca2cb8aa4430a9dd138e94 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:10:56,846 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(24283) have not registered within the timeout. The process is dead, probably it crashed during start.
(raylet, ip=10.159.0.81) [2022-08-24 09:10:57,552 E 24450 24450] core_worker.cc:137: Failed to register worker 1694186cf4e1328038f26330f7affebde5a2f8e3835b2e8652d4c9b1 to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.
(raylet, ip=10.159.0.81) [2022-08-24 09:11:56,850 E 22652 22652] (raylet) worker_pool.cc:502: Some workers of the worker process(24450) have not registered within the timeout. The process is dead, probably it crashed during start.
9 81
referrers 1
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
time:  561.3156089782715
```

**Comment by joshua-cogliati-inl** [comment]
This still exists in ray version 2.0.0.
```
$ pip list
Package            Version
------------------ ---------
aiosignal          1.2.0
attrs              22.1.0
certifi            2022.6.15
charset-normalizer 2.1.1
click              8.0.4
distlib            0.3.5
filelock           3.8.0
frozenlist         1.3.1
grpcio             1.43.0
idna               3.3
jsonschema         4.14.0
msgpack            1.0.4
numpy              1.23.2
pip                22.2.2
platformdirs       2.5.2
protobuf           3.20.1
pyrsistent         0.18.1
PyYAML             6.0
ray                2.0.0
requests           2.28.1
setuptools         65.3.0
six                1.16.0
urllib3            1.26.12
virtualenv         20.16.3
wheel              0.37.1
$ python --version             Python 3.10.6                                               
```
If I run:
```python
import socket
import time

l = []
for PORT in range(10002, 10010+1):
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.bind(('', PORT))
        s.listen(1)
        l.append(s)
    except OSError:
        print("unable to get ",PORT)

if len(l) > 0:
    input("Press enter to release ports:")
```
the ray_test2.py (with range bigger than the number of nodes) will fail to finish until I press enter.

Note that this is an artificial way to cause a real problem that we encounter.  When we are running production ray runs with hundreds of workers for multiple hours, the probability that one node will fail becomes sufficiently high that we pretty much always get this problem within a day. (We also encounter this problem both with and without using `--min-worker-port`)

**Comment by joshua-cogliati-inl** [comment]
A view in the dashboard. r1i2n5 is the one that is blocked:
![ray_failing](https://user-images.githubusercontent.com/7118148/190000298-320a94d2-4140-4026-9a63-3b4440eae4c8.png)

**Comment by stephanie-wang** [comment]
I'm afraid Ray is not expected to work if you specify too small of a range of worker ports and the ports are all in use. Your script will slow down quite a bit, and the errors about workers failing to start are expected, but eventually Ray should be timing out workers and retrying queued tasks on a different worker.

I took a look at the previous issues that you reported, and I think it's best that we try to solve the root problem there, since it seems like you are only using the worker ports as a workaround to limit how many workers run. I don't think it's a good idea to use this API for that purpose; it's only meant to be used in cases where the application can only expose certain ports, e.g., for security reasons, and Ray is not expected to work if you specify too narrow a range.

Will follow up on your other issue!

**Comment by joshua-cogliati-inl** [comment]
Hm, how long is it expected to take the workers to time out?  When we see this on our cluster in production runs, it seems like it can take over a day before the workers time out (or otherwise can run) and it starts running again (and then it hits the broken node, and gets blocked again).

I agree that one of the root problems is that there is no good api to limit the number of workers started (#27499)  I do think this is also a problem that one node being blocked from creating workers blocks all the nodes from finishing for excessively long times. (We have seen this with `--min-worker-port` and `--max-worker-port`, running out of memory, and hitting the ulimit for max user processes.)

Basically, we are hitting (in src/ray/core_worker/core_worker.cc):
```c++
  if (!raylet_client_status.ok()) {
    // Avoid using FATAL log or RAY_CHECK here because they may create a core dump file.
    RAY_LOG(ERROR) << "Failed to register worker " << worker_id << " to Raylet. "
                   << raylet_client_status;
    QuickExit();
  }
```
and somehow that does *not* result in the last few tasks getting assigned to idle workers.

They do get reassigned at the start (within seconds), but for the last few tasks, they can end up stuck on the broken node, and don't get reassigned for hours to days (and in production runs, it is possible that something else caused the ports, memory or processes needed to become available, so maybe they never get reassigned).

**Comment by stephanie-wang** [comment]
It should take <1min for the worker to time out.

From the other github issue, I think you're probably seeing tasks hanging because you have workers started by multiple drivers, and these can't be shared across the different drivers. So most likely, some workers have already taken up all of the available ports, and the scheduler can't create a suitable worker for the queued task.

We can fix Ray to reject a task if it fails to start any workers for it, but I think it will be tricky to fix this problem in all cases unless you modify your application. Fundamentally the worker ports are a shared and limited resource across the different drivers, and trying to always guarantee progress would require solving distributed deadlock.

**Comment by joshua-cogliati-inl** [comment]
I agree that the way we are using ray may be suboptimal, but the replication script here only uses one driver. From the Ray Architecture document: "Ray worker nodes are designed to be homogeneous, so that any single node may be lost without bringing down the entire cluster." Basically, this does fault injection on a single node, and that results in the entire cluster failing to make progress.

Bizarrely, if I want to finish say 10 tasks, (where 10 is > number of nodes I am running on) I can ask for 20 tasks (in this replication script), and it will finish the first 10, and then deadlock on the last few, so I could just ask for extra tasks to get ray to figure out how to do it, but that seems kind of wasteful. So I think ray already has solved the problem, it just has a bug somewhere that is breaking the last few tasks and not putting them on a different node.

**Comment by stephanie-wang** [comment]
I see, thanks! I've updated the issue title and we'll look into it, but I can't guarantee a timeline. So I strongly suggest that you find a way to work around using the min and max worker ports in the meantime.

**Comment by joshua-cogliati-inl** [comment]
Any hints as to where I should look in the ray code would be appreciated.

**Comment by joshua-cogliati-inl** [comment]
And this gives a segfault (regardless of if a node is blocked):
```python
import ray
import time
import gc
import threading
ray.init(address='auto')

start = time.time()

@ray.remote
def f(x):
    a = [2**2**2**2**2 for x in range(100000)]
    return x * x

@ray.remote
def s(x):
    a = [2**2**2**2**2 for x in range(10000)]
    return x * x

class StuffInQueue(threading.Thread):
    def run(self):
        count = 0
        num = 10
        while True:
            s_futures = [f.remote(i) for i in range(num)]
            print("started s futures", count, num)
            s_left = num
            while s_left > 0:
                for i in range(num):
                    try:
                        if s_futures[i] is not None:
                            runReturn = ray.get(s_futures[i], timeout=1e-10)
                            s_left -= 1
                            s_futures[i] = None
                    except ray.exceptions.GetTimeoutError:
                        pass
            print("got s futures", count)
            count += 1
            time.sleep(60.0)

print("starting stuff in queue")
stuff_in_queue = StuffInQueue()
stuff_in_queue.daemon = True
stuff_in_queue.start()
print("starting remote functions")
futures = [f.remote(i) for i in range(80)]
print("remotes started")
left = len(futures)
outputs = [None]*left
extra = 0
while left > 0:
    for i in range(len(futures)):
        if futures[i] is not None:
            try:
                fi = futures[i]
                runReturn = ray.get(fi, timeout=1e-10)
                print(i, runReturn)
                left += -1
                outputs[i] = runReturn
                futures[i] = None
                gc.collect()
                print("referrers", len(gc.get_referrers(fi)))
                del fi
                gc.collect()
            except ray.exceptions.GetTimeoutError:
                pass
    if left < 2 and extra > 0:
        print("adding extra")
        futures.append(f.remote(len(outputs)))
        outputs.append(None)
        left += 1
        extra += -1

print(outputs) # [0, 1, 4, 9]
end = time.time()
print("time: ",end - start)
```

```
[2022-09-19 13:31:31,661 C 37483 37515] core_worker_process.cc:29:  Check failed: !core_worker_process The process is already initialized for core worker.
*** StackTrace Information ***
/home/fred/miniconda3/envs/ray2/lib/python3.10/site-packages/ray/_raylet.so(+0xc3f93a) [0x2aaab304c93a] ray::operator<<()
/home/fred/miniconda3/envs/ray2/lib/python3.10/site-packages/ray/_raylet.so(+0xc41442) [0x2aaab304e442] ray::SpdLogMessage::Flush()
/home/fred/miniconda3/envs/ray2/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray6RayLogD1Ev+0x37) [0x2aaab304e757] ray::RayLog::~RayLog()
/home/fred/miniconda3/envs/ray2/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core17CoreWorkerProcess10InitializeERKNS0_17CoreWorkerOptionsE+0xb7) [0x2aaab2a05807] ray::core::CoreWorkerProcess::Initialize()
/home/fred/miniconda3/envs/ray2/lib/python3.10/site-packages/ray/_raylet.so(+0x4cb068) [0x2aaab28d8068] __pyx_pw_3ray_7_raylet_10CoreWorker_1__cinit__()
/home/fred/miniconda3/envs/ray2/lib/python3.10/site-packages/ray/_raylet.so(+0x4cbef3) [0x2aaab28d8ef3] __pyx_tp_new_3ray_7_raylet_CoreWorker()
python(_PyObject_MakeTpCall+0x192) [0x5555556925d2] _PyObject_MakeTpCall
python(_PyEval_EvalFrameDefault+0x5447) [0x55555568e0f7] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x1309) [0x555555689fb9] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x2ba1) [0x55555568b851] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x49d9) [0x55555568d689] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x1309) [0x555555689fb9] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x1309) [0x555555689fb9] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x6f5) [0x5555556893a5] _PyEval_EvalFrameDefault
python(_PyFunction_Vectorcall+0x6f) [0x555555699c2f] _PyFunction_Vectorcall
python(_PyEval_EvalFrameDefault+0x6f5) [0x5555556893a5] _PyEval_EvalFrameDefault
python(+0x1532d7) [0x5555556a72d7] method_vectorcall
python(+0x23f355) [0x555555793355] thread_run
python(+0x23f304) [0x555555793304] pythread_wrapper
/lib64/libpthread.so.0(+0x7ea5) [0x2aaaaacd6ea5] start_thread
/lib64/libc.so.6(clone+0x6d) [0x2aaaab8fab0d] clone
```

**Comment by stephanie-wang** [comment]
Sorry can you explain the context for these scripts? I tried running the latter on a local Ray instance but I'm not seeing the same segfault.
- what is the goal of the script?
- what is the setup for the Ray cluster? Was it started using `ray start` and if so, what is the command used?
- Can you provide the full output and preferably the logs from the run? You can use `ray logs` to zip the system logs.

**Comment by joshua-cogliati-inl** [comment]
The goal of the script was to force ray to finish by adding unnecessary work from a different thread similar to what I discussed in: https://github.com/ray-project/ray/issues/28071#issuecomment-1247030880

setup was as in https://github.com/ray-project/ray/issues/27499#issuecomment-1246894172  *without* running listen_range.py

**Comment by joshua-cogliati-inl** [comment]
If the script isn't doing anything too weird to be considered a bug, I can make a proper separate bug report on it.  As soon as I changed the script to stop running ray.get or f.remote in StuffInQueue after the main thread finished, it stopped throwing a segfault.

**Comment by stephanie-wang** [comment]
Hmm I see, most likely the segfault is harmless. It happens in multithreaded situations during teardown of the worker, if some C++ state gets torn down before all of the Python threads have joined. It doesn't look like we have an issue open for it, so feel free to do that, but it's relatively low priority since it's only at teardown.

**Comment by joshua-cogliati-inl** [comment]
So this code works (the extra thread manages to make ray able to finish):
```python
import ray
import time
import gc
import threading
ray.init(address='auto')

start = time.time()

@ray.remote
def f(x):
    a = [2**2**2**2**2 for x in range(100000)]
    return x * x

@ray.remote
def s(x):
    a = [2**2**2**2**2 for x in range(10000)]
    return x * x

class StuffInQueue(threading.Thread):
    def run(self):
        count = 0
        num = 10
        self.done = False
        while not self.done:
            s_futures = [f.remote(i) for i in range(num)]
            print("started s futures", count, num)
            s_left = num
            while s_left > 0 and not self.done:
                for i in range(num):
                    try:
                        if s_futures[i] is not None:
                            runReturn = ray.get(s_futures[i], timeout=1e-10)
                            s_left -= 1
                            s_futures[i] = None
                    except ray.exceptions.GetTimeoutError:
                        pass
            print("got s futures", count)
            count += 1
            time.sleep(60.0)

print("starting stuff in queue")
stuff_in_queue = StuffInQueue()
#stuff_in_queue.daemon = True
stuff_in_queue.start()
print("starting remote functions")
futures = [f.remote(i) for i in range(19)]
print("remotes started")
left = len(futures)
outputs = [None]*left
extra = 0
while left > 0:
    for i in range(len(futures)):
        if futures[i] is not None:
            try:
                fi = futures[i]
                runReturn = ray.get(fi, timeout=1e-10)
                print(i, runReturn)
                left += -1
                outputs[i] = runReturn
                futures[i] = None
                gc.collect()
                print("referrers", len(gc.get_referrers(fi)))
                del fi
                gc.collect()
            except ray.exceptions.GetTimeoutError:
                pass
    if left < 2 and extra > 0:
        print("adding extra")
        futures.append(f.remote(len(outputs)))
        outputs.append(None)
        left += 1
        extra += -1

print(outputs) # [0, 1, 4, 9]
end = time.time()
print("time: ",end - start)
stuff_in_queue.done = True
```

**Comment by joshua-cogliati-inl** [comment]
This works with a blocked node, but if I set extra to 0, it fails with a blocked node:
```python
import ray
import time
import gc
import threading
ray.init(address='auto')

start = time.time()

@ray.remote
def f(x):
    a = [2**2**2**2**2 for x in range(100000)]
    return x * x

print("starting remote functions")
futures = [f.remote(i) for i in range(19)]
print("remotes started")
left = len(futures)
needed_to_finish = len(futures)
outputs = [None]*left
#Run at most this number of extra tasks to get the main tasks to finish.
extra = 100
lowest_running = 0
while lowest_running  < needed_to_finish:
    lowest_running = len(futures)
    for i in range(len(futures)):
        if futures[i] is not None:
            if i < lowest_running:
                lowest_running = i
            try:
                fi = futures[i]
                runReturn = ray.get(fi, timeout=1e-10)
                print(i, runReturn, left)
                left += -1
                outputs[i] = runReturn
                futures[i] = None
                del fi
            except ray.exceptions.GetTimeoutError:
                pass
    if left < 10 and extra > 0:
        print("adding extra ", len(futures), lowest_running)
        futures.append(f.remote(len(outputs)))
        outputs.append(None)
        left += 1
        extra += -1

print(outputs) # [0, 1, 4, 9]
print(lowest_running)
end = time.time()
print("time: ",end - start)
```

**Comment by ericl** [comment]
Should we raise an error if the number of ports is too small for the CPU count? That way we can avoid inadvertent bricking of Ray via config, etc.

**Comment by joshua-cogliati-inl** [comment]
> Should we raise an error if the number of ports is too small for the CPU count? That way we can avoid inadvertent bricking of Ray via config, etc.

That would be useful, but it should be noted that the replication script sets `--num-cpus` to 1 and gives it 8 ports, and blocks the all the ports with listen in a different process on one of the nodes.  

Also, ray already produces this error message: `(raylet, ip=10.159.0.81) [2022-08-24 09:04:57,727 E 23321 23321] core_worker.cc:137: Failed to register worker d9bab59cebc6623b6888e5f9cba6756538880a24e452145b9b47bbbe to Raylet. Invalid: Invalid: No available ports. Please specify a wider port range using --min-worker-port and --max-worker-port.`

The reason I consider this a bug is because I only have to do this to one node, and that stops all the nodes from progressing.

**Comment by joshua-cogliati-inl** [comment]
This is still a problem in ray 2.1. I checked with python 3.10.8 and:
```
$ pip list
Package            Version
------------------ ---------
aiosignal          1.3.1
attrs              22.1.0
certifi            2022.12.7
charset-normalizer 2.1.1
click              8.0.4
distlib            0.3.6
filelock           3.8.2
frozenlist         1.3.3
grpcio             1.51.1
idna               3.4
jsonschema         4.17.3
msgpack            1.0.4
numpy              1.23.5
packaging          22.0
pip                22.3.1
platformdirs       2.6.0
protobuf           4.21.11
pyrsistent         0.19.2
PyYAML             6.0
ray                2.1.0
requests           2.28.1
setuptools         65.5.1
six                1.16.0
urllib3            1.26.13
virtualenv         20.17.1
wheel              0.38.4
```

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
